src/twisted/application/twist/_options.py

```diff
@@ -59,6 +59,10 @@ def __init__(self) -> None:
         self["reactorName"] = self.defaultReactorName
         self["logLevel"] = self.defaultLogLevel
         self["logFile"] = stdout
+        # An empty long description is explicitly set here as otherwise
+        # when executing from distributed trial twisted.python.usage will
+        # pull the description from `__main__` which is another entry point.
```

Comment:
It seems like the real condition that applies here is something other than "executing from distributed trial"?  Is it really *specifically* an interaction between disttrial and Options?  Or is it about `__main__` being set unexpected in that case?

Put another way, is `Options` broken when used with `python -m` in general or is it broken against disttrial?

Reply:
Is it about __main__ being a an unexpected in that case.

Then running with distributed worker, `__main__` is the distributed worker main and not the command main,

> Options broken when used with python -m in general or is it broken against disttrial?

Not sure if broken in general or just with distrial.

It works with `python -m` ... but not when called via disttrial

I would say that the automatic extraction of the long description from  `__main__` is broken in general... is just that it works with a direct invocation ffrom `bin/python -m `

-------------

This is hitting this branch
https://github.com/twisted/twisted/blob/7252b5e6d46ec5a834f8d79f6faba0c8252a36e3/src/twisted/python/usage.py#L538

Since the long description was not used anyway, I think it's best to explicitly set it here and don't rely on the automatic extraction of the long description.

Now... in the future, if a long description is added to the docstring of `twist` then this should be updated.

Reply:
I looked a little more closely and I think I understand this better now.

The test that hits this problem asks the `twist` options to describe themselves in two execution contexts.  (1) in the in-process context of the running test suite and (2) in a newly launched `python -m twisted` child process.

For (a), there are now two sub-cases we've identified.  (a) the test suite is being run by `trial ...` and (b) the test suite is being run by `trial -jN ...`.

For (1a), `__main__` is `twisted.scripts.trial` which has no docstring.  For (1b), `__main__` is `twisted.trial._disttrial.workertrial` which has a docstring.  For (2) `__main__` is `twisted.__main__` which has no docstring.

Thus, the test was (accidentally, I assume) correct because cases (1a) and (2) are in the same state with respect to their docstrings whereas case (1b) differs.

For now, it seems like the work-around is probably alright.  It only fixes one example of the general problem here - which is that `__main__.__doc__` doesn't seem to be particularly indicative of what an `Options` class is for.  Also, it's a bad API because it's action-at-a-distance, it's global state, and it's not even documented anywhere.

I think this behavior should be removed from `Options` and if folks want to set a longdesc, they should set a longdesc.

Do you agree?  Will you file a ticket for deprecation and removal?

Reply:
Yes. I agreed to deprecate the automatic long description feature in t.python.usage . I asked about this over IRC a few days ago.

https://twistedmatrix.com/trac/ticket/10133

---------------

Yes. Those are the 2 subclasses.

For (1a trial script) the test was not failing as twisted.scripts.trial._BasicOptions defined its own `longdesc` so the `__main__` docstring is not used at all.

https://github.com/twisted/twisted/blob/9a59de7abfe27f6c2ae1545374493f430938fb5b/src/twisted/scripts/trial.py#L165

This change should not be considered a workaround. It's an explicit definition for `longdesc` for `twist` CLI options.
For `twist` script I defined an empty `longdesc` as this was the automatically pulled value anyway.